From 5f658afd9a280a45a3f4a766dcfd4d5b38924216 Mon Sep 17 00:00:00 2001
From: David Gibson <dgibson@redhat.com>
Date: Mon, 15 Sep 2014 07:12:41 -0400
Subject: [virt] kvm/ppc: NULL return of kvmppc_mmu_hpte_cache_next should be handled

Message-id: <1410765214-16377-26-git-send-email-dgibson@redhat.com>
Patchwork-id: 94765
O-Subject: [PATCH 25/78] KVM: PPC: NULL return of kvmppc_mmu_hpte_cache_next should be handled
Bugzilla: 1123145 1123133 1123367
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1123145

Signed-off-by: David Gibson <dgibson@redhat.com>

commit 47d45d9f53a7c478fc83dff7b421cb4bc3ad9f94
Author: Zhouyi Zhou <zhouzhouyi@gmail.com>
Date:   Mon Dec 2 18:21:58 2013 +0800

    KVM: PPC: NULL return of kvmppc_mmu_hpte_cache_next should be handled

    NULL return of kvmppc_mmu_hpte_cache_next should be handled

    Signed-off-by: Zhouyi Zhou <yizhouzhou@ict.ac.cn>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_32_mmu_host.c b/arch/powerpc/kvm/book3s_32_mmu_host.c
index 7b65e17..678e753 100644
--- a/arch/powerpc/kvm/book3s_32_mmu_host.c
+++ b/arch/powerpc/kvm/book3s_32_mmu_host.c
@@ -243,6 +243,11 @@ next_pteg:
  /* Now tell our Shadow PTE code about the new page */
 
  pte = kvmppc_mmu_hpte_cache_next(vcpu);
+ if (!pte) {
+  kvm_release_pfn_clean(hpaddr >> PAGE_SHIFT);
+  r = -EAGAIN;
+  goto out;
+ }
 
  dprintk_mmu("KVM: %c%c Map 0x%llx: [%lx] 0x%llx (0x%llx) -> %lx\n",
       orig_pte->may_write ? 'w' : '-',
-- 
1.7.1